### DISCUSSION TITLE
Weapons Capabilities

### DISCUSSION THREAD

**Discussion by Timofffee** [Minestom/Minestom#1539] [Help]
Hey. I played around with the arena and I had a few questions:
1. There is no "rollback" of weapons in the Arena. This is similar to the old versions of minecraft. How can I get back the mechanics that allow me to "charge" a weapon if I wait for some time?
2. I have already partially found the answer to this question, but I would like it to be solved completely. How can a player's left and right clicks be tracked? I saw that the staff works through the animation of the hand, but if you click on a block with which you could interact, then this animation of the hand will work in the same way. Are there any alternatives?
3. Just for example, a stick is used as a staff. Is it possible, for example, to make a block of diamond used as a bow? in theory, of course, it is possible, but the question here is that the block on the right button will try to be put. Is it possible to change this somehow, or is this logic stored on the client side and it's better for me to use the bow variety?

**Comment by GreatWyrm** [comment]
Can't answer all questions, but for number 3:

Depends on what you mean by "used as a bow". If you simply want it to fire an arrow, yes, you can. If you want a draw animation (and slowdown) with a diamond block, you would have to make a resource pack with custom model data that switches a bow into a diamond block. Yes, it is possible to stop the block from being placed, you cancel the [PlayerBlockPlaceEvent](https://javadoc.minestom.net/net/minestom/server/event/player/PlayerBlockPlaceEvent.html).

**Reply by Timofffee** [reply]
Exactly. I'm sorry, my head hadn't had time to wake up yet when I asked this question.

**Reply by Tofaa2** [reply]
For question #2 The simplest and most efficient way imo would be to check the Animation packet for all swings, for block interactions you can use start Digging

**Reply by Timofffee** [reply]
About the first question. I found some solution, but it doesn't look particularly convenient. 
`player.getAttribute(Attribute.ATTACK_SPEED).setBaseValue(1.0f);`
Is it really necessary to change this parameter every time you switch?

**Reply by Tofaa2** [reply]
There is no cooldown on weapon swings, that's the way to show it client side but server side its the same

**Reply by Timofffee** [reply]
I just wanted to show the player that the weapon has a delay. It might not look so good for the staff that is in the arena. However, it will work well if you make weapons that cause some special damage if it is completely rolled back. It is clear that all this logic will have to be done independently.

**Comment by Timofffee** [comment]
Okay It seems I have to answer my own questions.

1. In order for the attack speed to be displayed on the client as filling in the sword scale, it is not enough for you to simply give the weapon an attribute. You need to track that this particular item was selected and then add its modifiers to the player ( `Player.getAttribute(Attribute.ATTACK_SPEED).addModifier(..)` ). The question that was asked earlier helped me partially in solving this problem: https://github.com/Minestom/Minestom/discussions/1089
2. The blow (left button) can almost always be tracked through the animation of the hand (but sometimes the interaction does the same thing, so you have to look for an additional solution here), but the interaction (right button) can be tracked through interaction with the block. Most of the time it should work.
3. In fact, it is possible, but not necessary. It is better to use a bow with a different level of damage. And yes, this will require the use of a resource pack: https://www.spigotmc.org/wiki/custom-item-models-in-1-9-and-up/

**Reply by GreatWyrm** [reply]
Minor addition to number 3, I'd recommend the use of custom model data instead, it's a lot easier to create than using damage values.
